Extend itemLimit config to KohaRest driver and ini file

alexklbuckley opened the pull request:
The itemLimit config is available with the Alma driver and it would be nice if the KohaRest driver could also use it. Especially, given the Koha-plugin-rest-di plugin will accept a 'limit' param.

To give more configurability we have enhanced the itemLimit config so that libraries can define different numbers of items to be displayed for different Koha biblio-level item types - which are defined in the 942$c subfield in Koha.

Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley committed:
Extend itemLimit config to KohaRest driver and ini file

The itemLimit config is available with the Alma driver and it would be
nice if the KohaRest driver could also use it. Especially, given the
Koha-plugin-rest-di plugin will accept a 'limit' param.

To give more configurability we have enhanced the itemLimit config so
that libraries can define different numbers of items to be displayed for
different Koha biblio-level item types - which are defined in the 942$c
subfield in Koha.

Sponsored-by: Auckland University of Technology, New Zealand

demiankatz commented on config/vufind/KohaRest.ini:
```diff
@@ -184,3 +184,14 @@ extraFields = item-issue:acceptTerms:pickUpLocation
+; By default, all items are displayed in the holdings tab, unless a rule has been defined 
+; for the relevant biblio-level item type for the biblio being loaded.
+[Holdings]
+;itemLimit[J] = 3
```
Might it make sense to add a default catch-all item limit, so users can set a single value for all bib levels if they wish? Maybe `itemLimit[DEFAULT]` or `itemLimit[*]` (if the .ini parser will let you get away with that, which it might not).

alexklbuckley replied:
thanks @demiankatz I've added that in my follow-up!

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -397,7 +397,14 @@ public function getStatuses($ids)
-        return $this->getItemStatusesForBiblio($id, $patron);
+        $biblio = $this->getBiblio($id);
+        $type   = $biblio['item_type'];
+        if (isset($this->config['Holdings']['itemLimit'][$type])) {
```
It's probably a good idea to initialize $options... and you could use null coalescing to avoid the if/else. I'd change this whole conditional block to something like:

```
$options = [
    'itemLimit' => $this->config['Holdings']['itemLimit'][$type]
        ?? $this->config['Holdings']['itemLimit']['DEFAULT']
        ?? ''
];
```

(Note that this also incorporates my suggestion above about a default fallback).

alexklbuckley replied:
thanks @demiankatz I've added that in my follow-up!

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -1971,23 +1978,37 @@ protected function getOAuth2Token($renew = false)
-                'errors' => true,
-            ]
-        );
+        if (isset($options['itemLimit']) && $options['itemLimit'] > 0) {
```
I would simplify this to:

```
        $requestParams = [
            'path' => [
                'v1', 'contrib', 'kohasuomi', 'availability', 'biblios', $id,
                'search',
            ],
            'errors' => true,
        ];
        if (($options['itemLimit'] ?? 0) > 0) {
            $requestParams['query'] => [ 'limit' => $options['itemLimit'] ];
        }
        $result = $this->makeRequest($requestParams);
```

alexklbuckley replied:
thanks @demiankatz I've added that in my follow-up!. I noticed the first => in             `$requestParams['query'] => [ 'limit' => $options['itemLimit'] ];` caused the unit tests to fail when I ran them locally with `vendor/bin/phing phpstan-console` so I changed it to a =

Then tested locally and the limit was being passed through to the Koha rest plugin successfully.

Please let me know if that change is incorrect?

demiankatz replied:
Yes, that was a typo, so your correction is correct. Sorry about that! The unit tests are doing their job! :-)

alexklbuckley replied:
Lovely, thank you!

demiankatz reviewed (changes requested):
Thanks, @alexklbuckley! A few small suggestions...

alexklbuckley committed:
Merge branch 'vufind-org:dev' into koharest-itemlimit

alexklbuckley committed:
Extend itemLimit config to KohaRest: Implement review feedback

- Add option of setting a single value for itemLimit that libraries can
  add a rule that works for all biblio-level item types: itemLimit[DEFAULT]

- Initialise and use null coalescing to avoid an if/else

- Simplify call to v1/contrib/kohasuomi/availability/biblios/$id/search
  passing in the limit param

Sponsored-by: Auckland University of Technology, New Zealand

demiankatz commented on config/vufind/KohaRest.ini:
```diff
@@ -184,3 +184,15 @@ extraFields = item-issue:acceptTerms:pickUpLocation
+; This section allows libraries to define different custom itemLimit rules for 
+; different biblio-level item types.
+; In Koha the biblio-level item type is defined in the 942$c subfield.
+; The string inside the [] is the Koha biblio-level item type. Set 'DEFAULT' to set a single value
```
Maybe this could say "Set 'DEFAULT' to set a fallback value that will be used for any item types not given a more specific setting." Or something like that?

alexklbuckley replied:
thanks @demiankatz ! Appreciate all your help.

I've added this wording change in my latest follow-up commit.

demiankatz reviewed (changes requested):
One small suggestion regarding the .ini file comments, but this otherwise looks good to me. Probably another one we should wait for @EreMaijala's final approval on, but I don't think it should be controversial. :-)

alexklbuckley committed:
Merge branch 'vufind-org:dev' into koharest-itemlimit

alexklbuckley committed:
Extend itemLimit config to KohaRest: Implementing review feedback

- Clarify wording in KohaRest.ini file

Sponsored-by: Auckland University of Technology, New Zealand

demiankatz commented:
Thanks, @alexklbuckley! Looks good to me, but I'll leave it unmerged until @EreMaijala gets back. One final question: is it intentional that the limit setting is only applied to getHoldings and not getStatus/getStatuses? I assume this _is_ intentional, since if we limited the status checks, it could result in inaccurate AJAX availability information... however, if performance is a concern here, that problem will be unchanged for status updates.

alexklbuckley commented:
hi @demiankatz thanks for that appreciate it! Yes, that is right. I have left off the limit setting for getStatus/getStatuses to prevent inaccurate AJAX availability information.

The purpose of this enhancement is performance improving, especially for journal bibs with hundreds of items attached. We have this working on the AUT VuFind dev site and it has greatly improved the load speeds of the journal bibs.

However, if you think not adding the limit setting for fetching statuses is going to be a problem, we would be happy to hear if you have any suggestions?

demiankatz commented:
> hi @demiankatz thanks for that appreciate it! Yes, that is right. I have left off the limit setting for getStatus/getStatuses to prevent inaccurate AJAX availability information.
> 
> The purpose of this enhancement is performance improving, especially for journal bibs with hundreds of items attached. We have this working on the AUT VuFind dev site and it has greatly improved the load speeds of the journal bibs.
> 
> However, if you think not adding the limit setting for fetching statuses is going to be a problem, we would be happy to hear if you have any suggestions?

If you're seeing improvement as it is, then I think it's okay. My question is, if you have a search result set containing a journal with many holdings, does that interfere with availability statuses for all of the records? My suspicion is that it will, but if that's the case, the work recently done in #3748 might actually help you work around it.

I've requested @EreMaijala's review to see if we're ready to merge this.

EreMaijala commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -397,7 +397,14 @@ public function getStatuses($ids)
-        return $this->getItemStatusesForBiblio($id, $patron);
+        $biblio = $this->getBiblio($id);
+        $type   = $biblio['item_type'];
+        $options = [
```
I'm a bit confused. Since we already support paging for holdings, why not add proper support for it here instead of overriding `$options`? Also, since module/VuFind/src/VuFind/ILS/Connection.php uses a setting with the same name, this is going to cause confusion later on as well.
So instead of handling the setting and changing `$options`, you could just check for `$options['itemLimit']` and  `$options['offset']` (see Alma::getHolding for example) and add the API parameters according to those. This would allow the user to page through the holdings instead of silently truncating the list.

If the itype-specific item limits are important, I'd add another setting like `itemLimitByType` and extend the `getConfig` method to check for it. If set, it would fetch the bib and copy any value found in itemLimitByType to the itemLimit setting to be returned from getConfig. Then itemLimit would still work as the default, and you'd only have to retrieve the bib if itemLimitByType is defined.

I realize that I might have explained myself really unclearly, so please ask if this just adds to the confusion. :)

alexklbuckley replied:
hi @EreMaijala thanks for your feedback.

I can certainly make those changes. one question i do have is because the koha-plugin-rest-di Biblio/Availability/Search->[_item_looper()](https://github.com/NatLibFi/koha-plugin-rest-di/blob/main/Koha/Plugin/Fi/KohaSuomi/DI/Koha/Biblio/Availability/Search.pm) does not accept the parameter of 'offset', I will need to add that to the koha-plugin-rest-di _item_looper()?

Thanks,
Alex

alexklbuckley replied:
thanks @EreMaijala , I've apart from my question about the Koha-plugin-rest-di plugin not accepting 'offset'. I think I have implemented all of your other points in my latest followup commit.

Could you please take a look?

Thank you very much.

EreMaijala replied:
Argh, you're right. We also need the total count of items to support paging. But it's a bit hairy since these affect the biblio availability status too. I don't really like this all, but oh well. I'll see if I can add the offset quickly. We just need to add a note in KohaRest.ini that to use the feature you'll need version X of the DI plugin...

EreMaijala replied:
Did you actually get the limit parameter to work either? As far as I can see it doesn't work in the DI plugin. The schema defines `limit_items`, but not for the biblio/availability/search endpoint.

alexklbuckley replied:
thanks for that @EreMaijala . 

Yes, I did get the limit parameter working. as the following screenshot shows I have 3 items returned for a journal biblio record due to VuFind making the following call to the DI plugin: 

![Screenshot from 2024-08-07 14-25-09](https://github.com/user-attachments/assets/7cb27826-85dd-4778-ac5b-c940ef64569d)

`2024-08-07T14:31:33+12:00 DEBUG (7): VuFind\ILS\Driver\KohaRest: [0.4933s] GET request <URL>/api/v1/contrib/kohasuomi/availability/biblios/1132644/search?limit=3`

alexklbuckley replied:
without that limit param then dozens more item records load.

EreMaijala replied:
Well, that's odd. I'm only getting 400 Bad Request with "Malformed query string" because `limit` is not defined as a valid parameter for the endpoint. I've now got it working locally, but your success is puzzling!

demiankatz replied:
Maybe worth comparing versions? Was something added and then removed or broken subsequently?

alexklbuckley committed:
Merge branch 'vufind-org:dev' into koharest-itemlimit

alexklbuckley committed:
Extend itemLimit config to KohaRest: Implementing review feedback

- Remove setting and changing $options
- Add itemLimitByType config to KohaRest.ini
- Fetch itemLimit and itemLimitByType from getConfig()

Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley commented:
> > hi @demiankatz thanks for that appreciate it! Yes, that is right. I have left off the limit setting for getStatus/getStatuses to prevent inaccurate AJAX availability information.
> > The purpose of this enhancement is performance improving, especially for journal bibs with hundreds of items attached. We have this working on the AUT VuFind dev site and it has greatly improved the load speeds of the journal bibs.
> > However, if you think not adding the limit setting for fetching statuses is going to be a problem, we would be happy to hear if you have any suggestions?
> 
> If you're seeing improvement as it is, then I think it's okay. My question is, if you have a search result set containing a journal with many holdings, does that interfere with availability statuses for all of the records? My suspicion is that it will, but if that's the case, the work recently done in #3748 might actually help you work around it.
> 
> I've requested @EreMaijala's review to see if we're ready to merge this.

Thanks for noting that @demiankatz ! 

Sorry, just to make sure I am understanding correctly. 

We have search result sets with many holdings, for example, the first bib in https://librarysearch.aut.ac.nz/vufind/Search/Results?lookfor=new+zealand+listener&type=AllFields . 

So what you're saying is when we upgrade AUT to 10.x (or if we backported https://github.com/vufind-org/vufind/pull/3748 ) we might be able to improve the performance of VuFind loading the holdings data in the search result page?

demiankatz commented:
> So what you're saying is when we upgrade AUT to 10.x (or if we backported #3748 ) we might be able to improve the performance of VuFind loading the holdings data in the search result page?

Yes, at least in a sense -- what #3748 adds is a setting to control the way the AJAX statuses are loaded. The default behavior is to request statuses in batches based on an intersection observer -- i.e. when the user scrolls some items into view, the status information for the visible items gets loaded in a batch. The problem with this approach is that if one item has a lot of items, it can slow down the loading of the entire batch. With these settings, you have the option of either forcing each status request to be made individually (so one is less likely to slow down others) or to request a single batch on page load (so the user doesn't have to wait to scroll to things to see them). Either of these approaches might yield better results in your situation, but it will all depend on your circumstances. It's at least something you can experiment with!

alexklbuckley commented:
> > So what you're saying is when we upgrade AUT to 10.x (or if we backported #3748 ) we might be able to improve the performance of VuFind loading the holdings data in the search result page?
> 
> Yes, at least in a sense -- what #3748 adds is a setting to control the way the AJAX statuses are loaded. The default behavior is to request statuses in batches based on an intersection observer -- i.e. when the user scrolls some items into view, the status information for the visible items gets loaded in a batch. The problem with this approach is that if one item has a lot of items, it can slow down the loading of the entire batch. With these settings, you have the option of either forcing each status request to be made individually (so one is less likely to slow down others) or to request a single batch on page load (so the user doesn't have to wait to scroll to things to see them). Either of these approaches might yield better results in your situation, but it will all depend on your circumstances. It's at least something you can experiment with!

Thanks so much for the explanation @demiankatz much appreciated! Yes, I will check in with our partner library if https://github.com/vufind-org/vufind/pull/3748 is something they would like us to backport, or if they are happy to wait until their upgrade to VuFind 10 to experiment with those options. 

Thanks again for all your help!

EreMaijala commented:
@alexklbuckley A [new release](https://github.com/NatLibFi/koha-plugin-rest-di/releases/tag/v23.11.08) of the DI plugin is now online. It has proper support for the `limit` and `offset` parameters for the `/v1/contrib/kohasuomi/availability/biblios/.../search` endpoint. It also includes the items_total field in the response, and this should be included in the response from getHolding method in VuFind like this:
```
return [
    'total' => $result['data']['items_total'], 
    'holdings' => [...]
];
```

Edit: Updated release link.

alexklbuckley committed:
Merge branch 'vufind-org:dev' into koharest-itemlimit

EreMaijala commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -1720,6 +1720,18 @@ public function getConfig($function, $params = [])
+            return [
+                'itemLimit' => $this->config['Holdings']['itemLimitByType'][$type]
+                ?? $this->config['Holdings']['itemLimit']
+                ?? '',
```
I think that null would be a more appropriate value here.

alexklbuckley replied:
Thanks @EreMaijala I've included that in my latest follow-up

EreMaijala reviewed (changes requested):
This just needs the few things mentioned in [my previous comment](https://github.com/vufind-org/vufind/pull/3839#issuecomment-2272999707).

alexklbuckley commented:
> @alexklbuckley A [new release](https://github.com/NatLibFi/koha-plugin-rest-di/releases/tag/v23.11.08) of the DI plugin is now online. It has proper support for the `limit` and `offset` parameters for the `/v1/contrib/kohasuomi/availability/biblios/.../search` endpoint. It also includes the items_total field in the response, and this should be included in the response from getHolding method in VuFind like this:
> 
> ```
> return [
>     'total' => $result['data']['items_total'], 
>     'holdings' => [...]
> ];
> ```
> 
> Edit: Updated release link.

thanks @EreMaijala ! I will try to get to adding my follow ups to this today!

alexklbuckley committed:
Merge branch 'dev' into koharest-itemlimit

alexklbuckley committed:
Extend itemLimit config to KohaRest: Implement review feedback

- Add 'offset' parameter to Koha REST plugin API call.

- Return null not '' from getConfig() if itemLimitByType and itemLimit
  are not set.

Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley commented:
> @alexklbuckley A [new release](https://github.com/NatLibFi/koha-plugin-rest-di/releases/tag/v23.11.08) of the DI plugin is now online. It has proper support for the `limit` and `offset` parameters for the `/v1/contrib/kohasuomi/availability/biblios/.../search` endpoint. It also includes the items_total field in the response, and this should be included in the response from getHolding method in VuFind like this:
> 
> ```
> return [
>     'total' => $result['data']['items_total'], 
>     'holdings' => [...]
> ];
> ```
> 
> Edit: Updated release link.

Thanks @EreMaijala I started working on this but got a bit stuck on where I need to add the above return. 

Would you mind please advising where it should be added in the KohaRest.php?

demiankatz reviewed (changes requested):
@alexklbuckley, I took a quick look to see if I could easily answer your question, but I see why you are a little confused. Hopefully @EreMaijala can clear it up when time permits. :-)

In the meantime, I see that there are some conflicts that need to be resolved here, likely because of #3856, which added some missing settings to KohaRest.ini. Do you mind resolving those and making sure your new settings end up in the right place? Please let us know if you run into any trouble with the conflict resolution!

alexklbuckley committed:
Merge branch 'dev' into koharest-itemlimit

alexklbuckley committed:
Extend itemLimit config to KohaRest: Implement review feedback

- Remove whitespace on the end of a line.

Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley commented:
> @alexklbuckley, I took a quick look to see if I could easily answer your question, but I see why you are a little confused. Hopefully @EreMaijala can clear it up when time permits. :-)
> 
> In the meantime, I see that there are some conflicts that need to be resolved here, likely because of #3856, which added some missing settings to KohaRest.ini. Do you mind resolving those and making sure your new settings end up in the right place? Please let us know if you run into any trouble with the conflict resolution!

Thanks @demiankatz !! I've fixed the conflicts and pushed that follow up and a further one removing a whitespace I inadvertently introduced :)

demiankatz committed:
Add end of file line break.

demiankatz committed:
Fix .ini merging.

demiankatz commented:
@alexklbuckley, there was a small problem here because there were two [Holdings] sections in KohaRest.ini, which would have prevented some settings from parsing correctly. I've merged them together and eliminated duplication; please take a look, confirm that the order is okay and that I didn't drop anything by accident, and make any further adjustments if necessary.

EreMaijala reviewed (changes requested):
@alexklbuckley You'll need to change getItemStatusesForBiblio to return the full array like:
`       ['total' => 5, 'holdings' => [...]];`
Then in getStatus and getStatuses just extract the `holdings` part from the array.

alexklbuckley committed:
Merge branch 'vufind-org:dev' into koharest-itemlimit

alexklbuckley committed:
Extend itemLimit config to KohaRest: Implement review feedback

- Return 'total' and 'holdings' from KohaRest.php getItemStatusesForBiblio() function.
- Make getStatus() and getStatuses() extract the holdings part of the array

Sponsored-by: Auckland University of Technology, New Zealand

alexklbuckley commented:
> @alexklbuckley, there was a small problem here because there were two [Holdings] sections in KohaRest.ini, which would have prevented some settings from parsing correctly. I've merged them together and eliminated duplication; please take a look, confirm that the order is okay and that I didn't drop anything by accident, and make any further adjustments if necessary.

Thanks @demiankatz ! I've taken a look and I think your merge is correct

alexklbuckley commented:
> @alexklbuckley You'll need to change getItemStatusesForBiblio to return the full array like: ` ['total' => 5, 'holdings' => [...]];` Then in getStatus and getStatuses just extract the `holdings` part from the array.

Thanks @EreMaijala ! I've just pushed a follow-up which hopefully does this correctly. Ready for you and @demiankatz and re-review

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -2048,6 +2056,11 @@ protected function getItemStatusesForBiblio($id, $patron = null, array $options
         }
 
+        // Get the total number of items returned from the API call and set it to
+        // a class variable. It is then used in VuFind\RecordTab\HoldingsILS for
```
I'm not sure what "set it to a class variable" refers to in this comment. You appear to be using a local array rather than a class property.

alexklbuckley replied:
Thanks @demiankatz I've included that fix in my follow up.

demiankatz reviewed (changes requested):
Apart from one question about a comment, I don't see any obvious problems here... but I'll wait for @EreMaijala's review since I can't easily test this myself.

EreMaijala commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -1744,6 +1746,18 @@ public function getConfig($function, $params = [])
                 'default_sort' => '+due_date',
             ];
+        } elseif ('Holdings' === $function) {
+            $limitByType = $this->config['Holdings']['itemLimitByType'];
```
This needs null coalescing (like ` ?? []`), otherwise this will cause an `Undefined array key` error if the setting isn't found.

alexklbuckley replied:
Thanks @EreMaijala this is included in my followup

EreMaijala commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -1744,6 +1746,18 @@ public function getConfig($function, $params = [])
+                $type   = $biblio['item_type'];
+            }
+            return [
+                'itemLimit' => $this->config['Holdings']['itemLimitByType'][$type]
```
Use `$limitByType` here

alexklbuckley replied:
Thanks @EreMaijala this is included in my followup

EreMaijala commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -1744,6 +1746,18 @@ public function getConfig($function, $params = [])
+            }
+            return [
+                'itemLimit' => $this->config['Holdings']['itemLimitByType'][$type]
+                ?? $this->config['Holdings']['itemLimit']
+                ?? null,
```
I'd indent these lines 4 more chars for clarity.

alexklbuckley replied:
Thanks @EreMaijala this is included in my followup

EreMaijala commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -2029,6 +2056,11 @@ protected function getItemStatusesForBiblio($id, $patron = null)
+        // the items paginator
+        $results['total'] = (int)$result['data']['items_total'];
+
         $statuses = [];
```
This variable is no longer used, right?

alexklbuckley replied:
Yes that's right. I've removed that line in my follow up

EreMaijala reviewed (changes requested):
This works, but needs just a few more tweaks. :)

alexklbuckley committed:
Merge branch 'vufind-org:dev' into koharest-itemlimit

alexklbuckley committed:
Extend itemLimit config to KohaRest: Implement review feedback

- Remove unused array
- Fix inaccurate comment
- Add null coalesce
- Improve indentation

Sponsored-by: Auckland University of Technology, New Zealand

EreMaijala commented on module/VuFind/src/VuFind/ILS/Driver/KohaRest.php:
```diff
@@ -2029,7 +2056,11 @@ protected function getItemStatusesForBiblio($id, $patron = null)
+        // Get the total number of items returned from the API call and save it to
+        // a local array. It is then used in VuFind\RecordTab\HoldingsILS for
+        // the items paginator
+        $results['total'] = (int)$result['data']['items_total'];
```
I'd avoid documenting functionality elsewhere in too much detail, because it could change later and cause confusion, so I suppose something like "Return total number of results for pagination." would be plenty enough here, but I won't hold this any longer because of that. :)

demiankatz replied:
I've gone ahead and made the suggested edit prior to merging this.

EreMaijala reviewed (approved):
Thanks, @alexklbuckley!

demiankatz committed:
Simplify comment.

demiankatz reviewed (approved):
Thanks, @alexklbuckley and @EreMaijala!

alexklbuckley commented:
Thanks as always @demiankatz and @EreMaijala !
